fhir-persistence/src/main/java/com/ibm/fhir/persistence/util/InputOutputByteStream.java

```diff
@@ -0,0 +1,285 @@
+        return new ByteInputStream();
+    }
+
+    /**
+     * Reset the offset to make the buffer appear empty. Does not change the current
+     * length (capacity). Note that any streams currently being used to read the data
+     * will end.
+     */
+    public void reset() {
+        this.offset = 0;
+    }
+
+    /**
+     * Simple performance comparison between ByteArrayOutputStream/ByteArrayInputStream and
+     * the comparable function provided by this class. Useful to run with a profiler.
+     * On: MacBook Pro 15-inch, 2016; 2.7 GHz Quad-Core Intel Core i7; 16 GB 2133 MHz LPDDR3
+     *     the performance improvement is around 15%.
+     * @param args
+     */
+    public static void main(String[] args) {
+        final int initialSize = 4096;
+        final int iters = 100000;
+        final int writeCount = 10000; // total about 40MiB
+
+        System.out.println("Starting micro-benchmark for " + InputOutputByteStream.class.getSimpleName() + ". This should take less that 1 minute on a modern CPU (2021)");
+
+        try {
+            // Just some random text to use as a filler for the buffer
+            final byte[] src = new String("{This is the story of...}").getBytes();
+            // Get a baseline with the original implementation
+            final long origStart = System.nanoTime();
+            for (int i=0; i<iters; i++) {
+                ByteArrayOutputStream bos = new ByteArrayOutputStream(initialSize);
+                for (int x=0; x<writeCount; x++) {
+                    bos.write(src);
+                }
+
+                ByteArrayInputStream bis = new ByteArrayInputStream(bos.toByteArray());
+                byte[] tmp = new byte[4096];
+                while (bis.read(tmp) != -1) {
+                    ; // NOP
+                }
+            }
+            final long origEnd = System.nanoTime();
+
+            // Now compare with the new approach
+            final long newStart = origEnd;
+            for (int i=0; i<iters; i++) {
+                InputOutputByteStream iobs = new InputOutputByteStream(initialSize);
+                OutputStream os = iobs.outputStream();
+                for (int x=0; x<writeCount; x++) {
+                    os.write(src);
+                }
+
+                InputStream is = iobs.inputStream();
+                byte[] tmp = new byte[4096];
+                while (is.read(tmp) != -1) {
+                    ; // NOP
+                }
+            }
+            final long newEnd = System.nanoTime();
+
+            double origSeconds = (origEnd - origStart) / 1e9;
+            double newSeconds = (newEnd - newStart) / 1e9;
+
+            System.out.println(String.format("Original = %5.3fs, New = %5.3fs, Percent improvement = %5.1f%%", origSeconds, newSeconds, 100.0 * (origSeconds - newSeconds) / origSeconds));
+        } catch (IOException x) {
+            System.out.println("Test failed: " + x.getMessage());
```

Comment:
let's move this into a test or tool under src/test/java

Reply:
OK I'll add a separate micro benchmark class for this in test. Agree this shouldn't be on the main class path.

Reply:
done